providers/implementations/signature/sm2sig.c

```diff
@@ -0,0 +1,502 @@
+    EVP_MD *md;
+    EVP_MD_CTX *mdctx;
+    size_t mdsize;
+} PROV_SM2_CTX;
+
+/* TODO: it seems SM2 doesn't need this */
+static int sm2sig_get_md_nid(const EVP_MD *md)
+{
+    /*
+     * Because the EC library deals with NIDs, we need to translate.
+     * We do so using EVP_MD_is_a(), and therefore need a name to NID
+     * map.
+     */
+    static const OSSL_ITEM name_to_nid[] = {
+        { NID_sm2,   OSSL_DIGEST_NAME_SM3 },
+    };
+    size_t i;
+    int mdnid = NID_undef;
+
+    if (md == NULL)
+        goto end;
+
+    for (i = 0; i < OSSL_NELEM(name_to_nid); i++) {
+        if (EVP_MD_is_a(md, name_to_nid[i].ptr)) {
+            mdnid = (int)name_to_nid[i].id;
+            break;
+        }
+    }
+
+    if (mdnid == NID_undef)
+        ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_DIGEST);
+
+ end:
+    return mdnid;
+}
+
+static void *sm2sig_newctx(void *provctx)
+{
+    PROV_SM2_CTX *psm2ctx = OPENSSL_zalloc(sizeof(PROV_SM2_CTX));
+
+    if (psm2ctx == NULL)
+        return NULL;
+
+    psm2ctx->libctx = PROV_LIBRARY_CONTEXT_OF(provctx);
+    /* don't allow to change MD, and in fact there is no such need */
+    psm2ctx->flag_allow_md = 0;
+    return psm2ctx;
+}
+
+static int sm2sig_setup_md(PROV_SM2_CTX *ctx,
+                           const char *mdname, const char *mdprops)
+{
+    if (mdname != NULL) {
+        EVP_MD *md = EVP_MD_fetch(ctx->libctx, mdname, mdprops);
+        int md_nid = sm2sig_get_md_nid(md);
+        size_t algorithmidentifier_len = 0;
+        const unsigned char *algorithmidentifier;
+
+        EVP_MD_free(ctx->md);
+        ctx->md = NULL;
+        ctx->mdname[0] = '\0';
+
+        algorithmidentifier =
+            sm2_algorithmidentifier_encoding(md_nid, &algorithmidentifier_len);
+
+        if (algorithmidentifier == NULL) {
+            EVP_MD_free(md);
+            return 0;
+        }
+
+        ctx->md = md;
+        OPENSSL_strlcpy(ctx->mdname, mdname, sizeof(ctx->mdname));
+        memcpy(ctx->aid, algorithmidentifier, algorithmidentifier_len);
+        ctx->aid_len = algorithmidentifier_len;
+    }
+    return 1;
+}
+
+static int sm2sig_signature_init(void *vpsm2ctx, void *ec)
+{
+    PROV_SM2_CTX *psm2ctx = (PROV_SM2_CTX *)vpsm2ctx;
+
+    if (psm2ctx == NULL || ec == NULL || !EC_KEY_up_ref(ec))
+        return 0;
+    EC_KEY_free(psm2ctx->ec);
+    psm2ctx->ec = ec;
+    return 1;
+}
+
+static int sm2sig_sign(void *vpsm2ctx, unsigned char *sig, size_t *siglen,
+                       size_t sigsize, const unsigned char *tbs, size_t tbslen)
+{
+    PROV_SM2_CTX *ctx = (PROV_SM2_CTX *)vpsm2ctx;
+    int ret;
+    unsigned int sltmp;
+    /* SM2 uses ECDSA_size as well */
+    size_t ecsize = ECDSA_size(ctx->ec);
+
+    if (sig == NULL) {
+        *siglen = ecsize;
+        return 1;
+    }
+
+    if (sigsize < (size_t)ecsize)
+        return 0;
+
+    if (ctx->mdsize != 0 && tbslen != ctx->mdsize)
+        return 0;
+
+    /* I would like to make the name something like SM2_sign... */
```

Comment:
Why?

Reply:
Because initially when I was working on this file, I chose the name `sm2_sign` here and I found it conflicted with the internal `sm2_sign` - so I changed the `sm2_sign` into `sm2sig_sign`. I am not sure if other SM2 provider implementor will also confuse with this and feel it not convenient.  (Other internal functions mostly are named with the upper-case name like ECDSA_sign or so)

Reply:
> Other internal functions mostly are named with the upper-case name like ECDSA_sign or so

ECDSA_sign is (unfortunately) not internal. Generally speaking we prefer to reserve the capitilized form for external functions and non-capitalized for internal.

Reply:
So, it seems the name similar to `sm2_internal_sign` is more reasonable

Reply:
I changed the name to `sm2_internal_sign` and `sm2_internal_verify` respectively